File: crates/better-auth-rs-utils/src/base32.ts
//inspired by oslo implementation by pilcrowonpaper: https://github.com/pilcrowonpaper/oslo/blob/main/src/encoding/base32.ts

import type { TypedArray, Uint8Array_ } from "./type";

/**
 * Returns the Base32 alphabet based on the encoding type.
 * @param hex - Whether to use the hexadecimal Base32 alphabet.
 * @returns The appropriate Base32 alphabet.
 */
function getAlphabet(hex: boolean): string {
	return hex
		? "0123456789ABCDEFGHIJKLMNOPQRSTUV"
		: "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
}

/**
 * Creates a decode map for the given alphabet.
 * @param alphabet - The Base32 alphabet.
 * @returns A map of characters to their corresponding values.
 */
function createDecodeMap(alphabet: string): Map<string, number> {
	const decodeMap = new Map<string, number>();
	for (let i = 0; i < alphabet.length; i++) {
		decodeMap.set(alphabet[i]!, i);
	}
	return decodeMap;
}

/**
 * Encodes a Uint8Array into a Base32 string.
 * @param data - The data to encode.
 * @param alphabet - The Base32 alphabet to use.
 * @param padding - Whether to include padding.
 * @returns The Base32 encoded string.
 */
function base32Encode(
	data: Uint8Array_,
	alphabet: string,
	padding: boolean,
): string {
	let result = "";
	let buffer = 0;
	let shift = 0;

	for (const byte of data) {
		buffer = (buffer << 8) | byte;
		shift += 8;
		while (shift >= 5) {
			shift -= 5;
			result += alphabet[(buffer >> shift) & 0x1f];
		}
	}

	if (shift > 0) {
		result += alphabet[(buffer << (5 - shift)) & 0x1f];
	}

	if (padding) {
		const padCount = (8 - (result.length % 8)) % 8;
		result += "=".repeat(padCount);
	}

	return result;
}

/**
 * Decodes a Base32 string into a Uint8Array.
 * @param data - The Base32 encoded string.
 * @param alphabet - The Base32 alphabet to use.
 * @returns The decoded Uint8Array.
 */
function base32Decode(data: string, alphabet: string): Uint8Array_ {
	const decodeMap = createDecodeMap(alphabet);
	const result: number[] = [];
	let buffer = 0;
	let bitsCollected = 0;

	for (const char of data) {
		if (char === "=") break;
		const value = decodeMap.get(char);
		if (value === undefined) {
			throw new Error(`Invalid Base32 character: ${char}`);
		}
		buffer = (buffer << 5) | value;
		bitsCollected += 5;

		while (bitsCollected >= 8) {
			bitsCollected -= 8;
			result.push((buffer >> bitsCollected) & 0xff);
		}
	}

	return Uint8Array.from(result);
}

/**
 * Base32 encoding and decoding utility.
 */
export const base32 = {
	/**
	 * Encodes data into a Base32 string.
	 * @param data - The data to encode (ArrayBuffer, TypedArray, or string).
	 * @param options - Encoding options.
	 * @returns The Base32 encoded string.
	 */
	encode(
		data: ArrayBuffer | TypedArray | string,
		options: { padding?: boolean } = {},
	): string {
		const alphabet = getAlphabet(false);
		const buffer =
			typeof data === "string"
				? new TextEncoder().encode(data)
				: new Uint8Array(data);
		return base32Encode(buffer, alphabet, options.padding ?? true);
	},

	/**
	 * Decodes a Base32 string into a Uint8Array.
	 * @param data - The Base32 encoded string or ArrayBuffer/TypedArray.
	 * @returns The decoded Uint8Array.
	 */
	decode(data: string | ArrayBuffer | TypedArray): Uint8Array_ {
		if (typeof data !== "string") {
			data = new TextDecoder().decode(data);
		}
		const alphabet = getAlphabet(false);
		return base32Decode(data, alphabet);
	},
};

/**
 * Base32hex encoding and decoding utility.
 */
export const base32hex = {
	/**
	 * Encodes data into a Base32hex string.
	 * @param data - The data to encode (ArrayBuffer, TypedArray, or string).
	 * @param options - Encoding options.
	 * @returns The Base32hex encoded string.
	 */
	encode(
		data: ArrayBuffer | TypedArray | string,
		options: { padding?: boolean } = {},
	): string {
		const alphabet = getAlphabet(true);
		const buffer =
			typeof data === "string"
				? new TextEncoder().encode(data)
				: new Uint8Array(data);
		return base32Encode(buffer, alphabet, options.padding ?? true);
	},

	/**
	 * Decodes a Base32hex string into a Uint8Array.
	 * @param data - The Base32hex encoded string.
	 * @returns The decoded Uint8Array.
	 */
	decode(data: string): Uint8Array_ {
		const alphabet = getAlphabet(true);
		return base32Decode(data, alphabet);
	},
};

File: crates/better-auth-rs-utils/src/base32.rs
//! Base32 / Base32hex (port of `base32.ts`, itself inspired by oslo).
//!
//! Hand-ported to match upstream byte-for-byte (and to avoid a same-named extern crate), mirroring
//! [`crate::base64`]. RFC 4648 §6 (`base32`, alphabet `A–Z2–7`) and §7 (`base32hex`, alphabet
//! `0–9A–V`). Encoding only — not a cryptographic primitive. Used by [`crate::otp`] to render the
//! `otpauth://` secret.

const STANDARD: &[u8; 32] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
const HEX: &[u8; 32] = b"0123456789ABCDEFGHIJKLMNOPQRSTUV";

fn encode_with(data: &[u8], alphabet: &[u8; 32], padding: bool) -> String {
    let mut out = String::with_capacity(data.len().div_ceil(5) * 8);
    // 5-bit packing; the buffer never holds more than 12 bits, so `u32` is ample.
    let mut buffer: u32 = 0;
    let mut shift: u32 = 0;
    for &byte in data {
        buffer = (buffer << 8) | u32::from(byte);
        shift += 8;
        while shift >= 5 {
            shift -= 5;
            out.push(alphabet[((buffer >> shift) & 0x1f) as usize] as char);
        }
    }
    if shift > 0 {
        out.push(alphabet[((buffer << (5 - shift)) & 0x1f) as usize] as char);
    }
    if padding {
        let pad = (8 - (out.len() % 8)) % 8;
        for _ in 0..pad {
            out.push('=');
        }
    }
    out
}

fn decode_with(data: &str, alphabet: &[u8; 32]) -> Result<Vec<u8>, Base32Error> {
    let mut reverse = [-1i16; 256];
    for (i, &c) in alphabet.iter().enumerate() {
        reverse[c as usize] = i as i16;
    }
    let mut out = Vec::with_capacity(data.len() / 8 * 5);
    let mut buffer: u32 = 0;
    let mut bits: u32 = 0;
    for ch in data.chars() {
        if ch == '=' {
            break;
        }
        let code = ch as u32;
        let value = if code < 256 {
            reverse[code as usize]
        } else {
            -1
        };
        if value < 0 {
            return Err(Base32Error::InvalidCharacter(ch));
        }
        buffer = (buffer << 5) | value as u32;
        bits += 5;
        if bits >= 8 {
            bits -= 8;
            out.push(((buffer >> bits) & 0xff) as u8);
        }
    }
    Ok(out)
}

/// Standard base32 (RFC 4648 §6, `A–Z2–7` alphabet).
#[allow(clippy::module_inception)]
pub mod base32 {
    use super::{Base32Error, STANDARD, decode_with, encode_with};

    /// Encode bytes as standard base32. `padding` controls trailing `=`.
    #[must_use]
    pub fn encode(data: impl AsRef<[u8]>, padding: bool) -> String {
        encode_with(data.as_ref(), STANDARD, padding)
    }

    /// Decode a standard base32 string (stops at the first `=`).
    pub fn decode(data: &str) -> Result<Vec<u8>, Base32Error> {
        decode_with(data, STANDARD)
    }
}

/// Extended-hex base32 (RFC 4648 §7, `0–9A–V` alphabet).
pub mod base32hex {
    use super::{Base32Error, HEX, decode_with, encode_with};

    /// Encode bytes as base32hex. `padding` controls trailing `=`.
    #[must_use]
    pub fn encode(data: impl AsRef<[u8]>, padding: bool) -> String {
        encode_with(data.as_ref(), HEX, padding)
    }

    /// Decode a base32hex string (stops at the first `=`).
    pub fn decode(data: &str) -> Result<Vec<u8>, Base32Error> {
        decode_with(data, HEX)
    }
}

/// Error returned when decoding encounters a character outside the active alphabet.
#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
pub enum Base32Error {
    /// A non-alphabet, non-padding character was encountered (upstream throws
    /// `Invalid Base32 character: {0}`).
    #[error("Invalid Base32 character: {0}")]
    InvalidCharacter(char),
}

#[cfg(test)]
#[path = "base32.test.rs"]
mod base32_tests;
